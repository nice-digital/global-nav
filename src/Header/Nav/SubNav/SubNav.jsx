import React from "react";
import cksIcon from "./images/cks.svg";
import bnfIcon from "./images/bnf.svg";
import bnfcIcon from "./images/bnfc.svg";
import { getCallbackFunction } from "../../../utils";
import PropTypes from "prop-types";

const images = {
	cks: cksIcon,
	bnf: bnfIcon,
	bnfc: bnfcIcon,
};

import {
	trackEvent,
	defaultEventCategory,
	headerClickEventAction,
} from "../../../tracker";

import styles from "./SubNav.module.scss";

const SubNav = (props) => {
	const handleClick = (e) => {
		e.preventDefault();

		const { currentTarget } = e;
		const href = currentTarget.getAttribute("href");

		const eventLabel =
			currentTarget.textContent ||
			currentTarget.innerText ||
			currentTarget.ariaLabel;

		trackEvent(
			defaultEventCategory,
			headerClickEventAction,
			eventLabel,
			null,
			href,
			function () {
				const { onNavigating } = props;

				const onNavigatingCallback = getCallbackFunction(onNavigating);

				if (onNavigatingCallback) {
					onNavigatingCallback({
						element: currentTarget,
						href: href,
					});
				} else window.location.assign(href);
			}
		);
	};

	return (
		<div className={styles.wrapper}>
			<ul className={styles.list} aria-label={`${props.text} links`}>
				{props.links.map(function (subLink, i) {
					let ariaCurrent = null;

					if (typeof window !== "undefined") {
						if (window.location.pathname === subLink.href) {
							ariaCurrent = "page";
						} else if (window.location.pathname.indexOf(subLink.href) === 0)
							ariaCurrent = true;
					}

					return (
						<li key={i} className={subLink.image && styles.imageLink}>
							<a
								href={subLink.href}
								aria-current={ariaCurrent}
								className={styles.link}
								onClick={handleClick}
								aria-label={subLink.text}
							>
								{subLink.image ? (
									<img
										src={images[subLink.image]}
										className={styles.image}
										alt=""
									/>
								) : (
									subLink.text
								)}
							</a>
						</li>
					);
				})}
			</ul>
		</div>
	);
};

export default SubNav;

//TODO Convert proptypes to typescript
SubNav.propTypes = {
	text: PropTypes.string.isRequired,
	links: PropTypes.arrayOf(
		PropTypes.shape({
			href: PropTypes.string.isRequired,
			text: PropTypes.string.isRequired,
		})
	),
	onNavigating: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
};
