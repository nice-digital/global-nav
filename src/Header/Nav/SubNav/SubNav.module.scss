@use '../../../styles/global-nav-colours' as *;
@use '@nice-digital/nds-core/scss/colours';
@use '@nice-digital/nds-core/scss/layout';
@use '@nice-digital/nds-core/scss/media-queries';
@use '@nice-digital/nds-core/scss/spacing';
@use '@nice-digital/nds-core/scss/utils';
@use '../../header-settings';

.wrapper {
  background: $subnav-background;
  color: $subnav-text;
  display: block;
}

.list {
  margin: 0;
  padding: 0;

  li {
    list-style: none;
    margin: 0;
  }
}

.link {
  border: utils.rem(spacing.$x-small) solid $subnav-link-border;
  border-bottom-width: 0;
  border-top-width: 0;
  color: colours.$text;
  display: block;
  font-weight: normal;
  padding: utils.rem(
    spacing.$small spacing.$medium spacing.$small spacing.$large
  );
  text-decoration: none;

  &:focus {
    color: colours.$text;
    outline: utils.rem(spacing.$x-small) solid colours.$focus;
    outline-offset: utils.rem(-(spacing.$small));
  }

  &:visited,
  &:hover,
  &:active {
    border: utils.rem(spacing.$x-small) solid $subnav-link-hover-border;
    border-bottom-width: 0;
    border-top-width: 0;
    color: $subnav-link-hover-text;
  }

  &:hover {
    background: $subnav-link-hover-background;
    color: colours.$text;
    text-decoration-line: underline;
  }
}

.imageLink {
  display: none;
}

@include media-queries.mq($from: md) {
  .wrapper {
    border-bottom: utils.rem(spacing.$xx-small) solid $subnav-border;
    bottom: 0;
    height: utils.rem(header-settings.$sub-nav-height);
    left: 0;
    position: absolute;
    width: 100%;
  }

  .imageLink {
    display: inline-block;
  }

  .list {
    @include layout.container;
    height: 100%;

    li {
      display: inline-block;
      height: 100%;
      vertical-align: middle;
    }
  }

  .link {
    border: 0;
    display: block;
    height: 100%;
    padding: utils.rem(0 spacing.$medium);

    &[aria-current='true'],
    &[aria-current='page'] {
      background: transparent;

      &:after {
        display: none;
      }
    }

    // Use pseudo to centrally align text
    &:before {
      content: '';
      display: inline-block;
      height: 100%;
      vertical-align: middle;
    }

    &:visited,
    &:hover,
    &:active {
      border: 0;
    }
  }
}

.image {
  height: 100%;
}
