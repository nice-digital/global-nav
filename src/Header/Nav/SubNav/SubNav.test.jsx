import React from "react";
import SubNav from "./SubNav";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

describe("SubNav", () => {
	const links = [
		{
			text: "Home",
			href: "/",
		},
		{
			text: "Drugs",
			href: "/drugs/",
		},
		{
			text: "Interactions",
			href: "/interactions/",
		},
	];

	it("Renders without crashing", () => {
		const wrapper = shallow(<SubNav text="BNF" links={links} />);
		expect(wrapper).toHaveLength(1);
	});

	it("Matches snapshot", () => {
		const wrapper = shallow(<SubNav text="BNF" links={links} />);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it("Adds aria-current=page attribute to link when matches current URL", () => {
		window.location.pathname = links[1].href;
		const wrapper = shallow(<SubNav text="BNF" links={links} />);

		expect(wrapper.find("a").at(1).props()["aria-current"]).toEqual("page");
	});

	it("Adds aria-current=true attribute to link when partially matches current URL", () => {
		window.location.pathname = links[1].href + "abacavir.html";
		const wrapper = shallow(<SubNav text="BNF" links={links} />);

		expect(wrapper.find("a").at(0).props()["aria-current"]).toEqual(true);
	});
});
