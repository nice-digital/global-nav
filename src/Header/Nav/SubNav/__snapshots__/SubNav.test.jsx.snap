// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubNav Matches snapshot 1`] = `
<div
  className="wrapper"
>
  <ul
    aria-label="BNF links"
    className="list"
  >
    <li
      key="0"
    >
      <a
        aria-current="page"
        aria-label="Home"
        className="link"
        href="/"
      >
        Home
      </a>
    </li>
    <li
      key="1"
    >
      <a
        aria-current={null}
        aria-label="Drugs"
        className="link"
        href="/drugs/"
      >
        Drugs
      </a>
    </li>
    <li
      key="2"
    >
      <a
        aria-current={null}
        aria-label="Interactions"
        className="link"
        href="/interactions/"
      >
        Interactions
      </a>
    </li>
  </ul>
</div>
`;
